rpi5.yaml

```diff
@@ -271,25 +271,7 @@ images:
         size: 512 MiB
         items:
           "armstub8-2712.bin": "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/armstub8-2712.bin"
-          "bootcode.bin": "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/bootcode.bin"
-          "config.txt":   "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/config.txt"
-          "cmdline.txt":  "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/cmdline.txt"
-          "fixup.dat":    "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/fixup.dat"
-          "fixup_cd.dat": "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/fixup_cd.dat"
-          "fixup_db.dat": "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/fixup_db.dat"
-          "fixup_x.dat":  "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/fixup_x.dat"
-          "fixup4.dat":   "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/fixup4.dat"
-          "fixup4cd.dat": "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/fixup4cd.dat"
-          "fixup4db.dat": "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/fixup4db.dat"
-          "fixup4x.dat":  "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/fixup4x.dat"
-          "start.elf":    "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/start.elf"
-          "start_cd.elf": "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/start_cd.elf"
-          "start_db.elf": "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/start_db.elf"
-          "start_x.elf":  "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/start_x.elf"
-          "start4.elf":   "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/start4.elf"
-          "start4db.elf": "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/start4db.elf"
-          "start4cd.elf": "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/start4cd.elf"
-          "start4x.elf":  "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/start4x.elf"
+          "/": "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles"
```

Comment:
Have you tested this? Does ninja correctly handles changes in the files?

Reply:
ninja has nothing to do with this - this is handled by **Rouge** tool. And it works for directory.

Reply:
`ninja` has everything to do with this, as it parses dependencies and decides what should be done before invoking the `rouge` tool. With your approach there is no guarantee that image file will include any bootfiles at all. If Yocto for some reason will not build them, rouge will pack just an empty directory. Also, ninja might not trigger image rebuild if some files inside directory were changed. We need to test this.

Reply:
Have you tried by your self delete any file deploy/images/? run **ninja** it will succeed, but **ninja full.img** - will fail. So do not know what "ninja handles dependency" function you refer here to :(

Next in "build.ninja" there is below - which is just call of **rouge** with parameters:
build full.img: rouge yocto/build-domd/tmp/deploy/images/raspberrypi5/armstub8-2712.bin $
    yocto/build-domd/tmp/deploy/images/raspberrypi5/bootfiles $
    yocto/build-domd/tmp/deploy/images/raspberrypi5/Image.gz $


And the last, and most important, we DO NOT maintain **bootfiles** recipe. it come from **meta-raspberrypi** and can be changed there by adding or removing some bootloader files. Now if smth is removed - ok **rouge** will fail. But if added - the build will succeed (including image generation), but RPI5 boot will fail. It make no sense for us to maintain full list of RPI5 bootfiles here (and moreover It is not safe) as we do not own it. What is reasonable is to have dependency on one or two files and if it's satisfied then assume **bootfiles** build is ok and just copy ALL bootfiles in the rootfs image.

Reply:
And moulin documentation:
https://moulin.readthedocs.io/en/latest/user-reference.html
 yocto builder https://moulin.readthedocs.io/en/latest/user-reference.html#yocto-builder

`target_images - list of image files that should be generated by this component as a result of invoking $ bitbake {build_target}. Every component should generate at least one image file.
`
Note. **Every component should generate at least one image file.** So no need to list ALL binaries.

`additional_deps - list of additional dependencies. This is basically target_images produced by other components. You can use those to implement build dependencies between components. For example, if your system needs to have DomU’s kernel image on Dom0 file system, you might want to add path to DomU’s kernel into additional_deps of Dom0’s config. This will ensure that Dom0 will be built after DomU.`

Note. **additional_deps** description clearly mention that NO NEED to add dependencies for ALL images, just  for key images.

Rouge docs: https://moulin.readthedocs.io/en/latest/rouge.html#filesystem-image-with-files   
`
You can specify parent folders for remote and these folders will be created on the destination filesystem. You may specify not only files but directories also. If the local directory contains subdirectories, they will be created unders the remote directory. 
`
As per above, just again. this change is valid usage of moulin/rouge.

Reply:
>  Now if smth is removed - ok rouge will fail. But if added - the build will succeed (including image generation), but RPI5 boot will fail. It make no sense for us to maintain full list of RPI5 bootfiles here (and moreover It is not safe) as we do not own it. 

In my opinion it is better to fail during the build process, not during runtime. This will save us a lot of headache in case of any issues in the future. This is another reason why I want to maintain the complete list of required files.